scintillometry/utils/phase_utils.py

```diff
@@ -0,0 +1,135 @@
+       obs_freq : float or `~astropy.units.Quantity`.
+           Observing frequency default units is 'MHz'
+       ephem : str
+           The solar system ephemeris version.
+       bipm_version : str
+           The BIPM clock correction version.
+
+       Note
+       ----
+       This method provides high precision phase calculation. The fractional
+       phase is between -0.5 to 0.5.
+    """
+    def __init__(self, par_file, obs, obs_freq, ephem='de436', include_bipm=True,
+                 bipm_version='BIPM2015', include_gps=True, planets=False,
+                 tdb_method="default", control_from_par=False, **kwargs):
+        self.par_file = par_file
+        self.model = get_model(self.par_file)
+        self.toa_maker = PintToas(ephem=ephem, include_bipm=include_bipm,
+                                  bipm_version=bipm_version,
+                                  include_gps=include_gps,
+                                  planets=planets, tdb_method=tdb_method,
+                                  **kwargs)
+        self.obs = obs
+        self.obs_freq = obs_freq
+
+    def __call__(self, t):
+        """Compute the apparent phase at one or more time stemp
+
+           Parameter
+           ---------
+           t : `astropy.time.Time` object
+               The input time stemps.
+
+           Return
+           ------
+           The apparent pulse phase at time t. The phases are return as
+           'integer phase, fractional phase'
+        """
+        toas = self._make_toas(t)
+        ph = self.model.phase(toas)
+        return u.Quantity((ph.int, ph.frac), u.cycle)
+
+    def _make_toas(self, t):
+        toa_list = make_toa_list(t, self.obs, self.obs_freq)
+        toas = self.toa_maker.make_toas(toa_list=toa_list)
+        return toas
+
+    def apparent_spin_freq(self, t):
+        """Compute the apparent spin frequency at one or more time stemps.
+
+            Parameter
+            ---------
+            t : `astropy.time.Time` object
+                The input time stemps.
+
+            Return
+            ------
+            The apparent spin frequency at time t.
+        """
+        toas = self._make_toas(t)
+        apprnt_f0 = self.model.d_phase_d_toa(toas)
+        return apprnt_f0
+
+
+class PolycoPhase(object):
+    """A utility class for a Tempo style polyco phase calculation.
+
+       Parameter
+       ---------
+       polyco_file : str
+           Tempo style polyco file.
+    """
+    def __init__(self, polyco_file):
+        self.polyco = Polyco(polyco_file)
+
+    def __call__(self, t):
+        """Compute the apparent phase at one or more time stemp
+
+           Parameter
+           ---------
+           t : `astropy.time.Time` object
+               The input time stemps.
+
+           Return
+           ------
+           The apparent pulse phase at time t. The phases are return as
+           'integer phase, fractional phase'.
+        """
+        t = t._apply(np.atleast_1d)
```

Comment:
Is the 1-D time needed for `Polyco`?

Reply:
One may want to see the phase only at one time. This line just makes sure the input time is iteratable.

Reply:
OK. Though I think `Polyco` should handle that natively... Anyway, can be done in another PR.

Reply:
Are you sure there is actually a problem with 0-D times? Is it in `Polyco` or somewhere else? I don't know that anything should assume an iterable.